macOS shebang launcher (SHEBANG_TPL_MACOS) cd's away from caller cwd before exec'ing python, breaking relative paths

### What happened?

On macOS, relenv's polyglot shell+python launcher (`SHEBANG_TPL_MACOS` in `relenv/common.py`) `cd`s into the script's own (post-symlink-resolution) directory while resolving `REALPATH`, and never returns to the caller's directory before `exec`-ing the bundled python. Python therefore inherits a cwd of the install dir (e.g. `/opt/salt`), not the directory the user invoked the command from.

Current template:

```python
SHEBANG_TPL_MACOS = textwrap.dedent(
    """\
#!/bin/sh
"true" ''''
TARGET_FILE=$0
cd "$(dirname "$TARGET_FILE")" || return
TARGET_FILE=$(basename "$TARGET_FILE")
# Iterate down a (possible) chain of symlinks
while [ -L "$TARGET_FILE" ]
do
    TARGET_FILE=$(readlink "$TARGET_FILE")
    cd "$(dirname "$TARGET_FILE")" || return
    TARGET_FILE=$(basename "$TARGET_FILE")
done
PHYS_DIR=$(pwd -P)
REALPATH=$PHYS_DIR/$TARGET_FILE
# shellcheck disable=SC2093
"exec" "$(dirname "$REALPATH")"{} "$REALPATH" "$@"
' '''
"""
)
```

Note this is macOS-only: `SHEBANG_TPL_LINUX` uses `readlink -f` in a single command substitution and never `cd`s, so the caller's cwd is preserved on Linux. The macOS template emulates `readlink -f` with a `cd`/`readlink` loop (since macOS lacks `readlink -f` pre-Big Sur era coreutils) but forgets to restore the original directory afterwards.

### Real-world impact

Every entry point of the official Salt macOS onedir (`salt`, `salt-ssh`, `salt-call`, `salt-run`, ...) is generated from this template. Any relative path Salt resolves against `os.getcwd()` breaks: `Saltfile` auto-discovery from the current directory, `config_dir: ./...` inside a Saltfile, `--config-dir=./...`, `root_dir: ./` in master config, `ssh_pre_flight: ./...` in a salt-ssh roster — all resolve against `/opt/salt` instead of the user's actual cwd.

Originally reported (with full reproducer against `com.saltstack.salt` 3007.14 / relenv 0.22.4) at saltstack/salt#69027 and re-filed here at the Salt maintainers' request, since the launcher template is generated by relenv and only consumed by Salt at install time.

### Reproducer

```sh
$ mkdir -p /tmp/saltrepro && cd /tmp/saltrepro
$ cat > Saltfile <<'YAML'
salt-ssh:
  config_dir: ./salt_root
YAML
$ mkdir salt_root
$ cat > salt_root/master <<'YAML'
root_dir: ./
file_roots:
  base:
    - ./states
YAML
$ salt-ssh -l debug '*' test.ping
[DEBUG   ] Missing configuration file: /opt/salt/salt_root/master
```

`./salt_root` resolved against `/opt/salt`, not `/tmp/saltrepro`.

### Expected behavior

Python's cwd at entry should be the directory the user invoked the command from, matching the Linux template's behavior. Concretely: capture `pwd` before the `cd`/`readlink` loop and `cd` back before the `exec`.

### Suggested patch

```python
SHEBANG_TPL_MACOS = textwrap.dedent(
    """\
#!/bin/sh
"true" ''''
ORIG_CWD=$(pwd)
TARGET_FILE=$0
cd "$(dirname "$TARGET_FILE")" || return
TARGET_FILE=$(basename "$TARGET_FILE")
# Iterate down a (possible) chain of symlinks
while [ -L "$TARGET_FILE" ]
do
    TARGET_FILE=$(readlink "$TARGET_FILE")
    cd "$(dirname "$TARGET_FILE")" || return
    TARGET_FILE=$(basename "$TARGET_FILE")
done
PHYS_DIR=$(pwd -P)
REALPATH=$PHYS_DIR/$TARGET_FILE
cd "$ORIG_CWD" || return
# shellcheck disable=SC2093
"exec" "$(dirname "$REALPATH")"{} "$REALPATH" "$@"
' '''
"""
)
```

### Workaround

Invoke the bundled python on the launcher script directly from the desired cwd; the bash blob lives inside a Python triple-quoted string and is ignored when python parses the file:

```sh
cd "$REPO_ROOT" && exec /opt/salt/bin/python3.10 /opt/salt/salt-ssh "$@"
```

### Versions

- relenv 0.22.4 (as shipped in Salt 3007.14 official macOS `.pkg`)
- macOS 15 / Darwin 25.4.0, Apple Silicon

--- a/relenv/common.py
+++ b/relenv/common.py
@@ -156,18 +156,26 @@ SHEBANG_TPL_MACOS = textwrap.dedent(
     """\
 #!/bin/sh
 "true" ''''
-TARGET_FILE=$0
-cd "$(dirname "$TARGET_FILE")" || return
-TARGET_FILE=$(basename "$TARGET_FILE")
-# Iterate down a (possible) chain of symlinks
-while [ -L "$TARGET_FILE" ]
-do
-    TARGET_FILE=$(readlink "$TARGET_FILE")
-    cd "$(dirname "$TARGET_FILE")" || return
+# Resolve $0 to its physical path inside a subshell so the cd's required to
+# walk symlinks on macOS (which lacks `readlink -f` in /usr/bin pre-Big Sur)
+# don't leak out and clobber the caller's working directory.  Without the
+# subshell, every entry point produced by relenv on macOS started Python
+# with cwd set to the install dir, breaking anything that resolved relative
+# paths against os.getcwd() (Saltfile auto-discovery, --config-dir=./...,
+# config_dir: ./..., etc).  See
+# https://github.com/saltstack/relenv/issues/293.
+REALPATH=$(
+    TARGET_FILE=$0
+    cd "$(dirname "$TARGET_FILE")" || exit
     TARGET_FILE=$(basename "$TARGET_FILE")
-done
-PHYS_DIR=$(pwd -P)
-REALPATH=$PHYS_DIR/$TARGET_FILE
+    while [ -L "$TARGET_FILE" ]
+    do
+        TARGET_FILE=$(readlink "$TARGET_FILE")
+        cd "$(dirname "$TARGET_FILE")" || exit
+        TARGET_FILE=$(basename "$TARGET_FILE")
+    done
+    echo "$(pwd -P)/$TARGET_FILE"
+)
 # shellcheck disable=SC2093
 "exec" "$(dirname "$REALPATH")"{} "$REALPATH" "$@"
 ' '''

--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -316,6 +316,56 @@ def test_format_shebang_newline() -> None:
     assert format_shebang("python3", SHEBANG_TPL_LINUX).endswith("\n")
 
 
+@mark_skipif(sys.platform == "win32", reason="POSIX shell launcher; no /bin/sh on Windows")
+def test_macos_shebang_preserves_caller_cwd(tmp_path: pathlib.Path) -> None:
+    """
+    Regression for https://github.com/saltstack/relenv/issues/293.
+
+    The pre-fix SHEBANG_TPL_MACOS resolved $0 through a cd / readlink loop
+    that ran in the script's own shell, so by the time it `exec`'d Python
+    the caller's working directory had been replaced with the install
+    directory.  The new template confines those cd's to a subshell.
+
+    Exercised here on plain /bin/sh — Python's launcher polyglot is
+    POSIX-portable, so a Linux runner reproduces the bug the same way a
+    macOS one does.
+    """
+    install_dir = tmp_path / "install"
+    shim_dir = tmp_path / "shim"
+    caller_dir = tmp_path / "caller"
+    install_dir.mkdir()
+    shim_dir.mkdir()
+    caller_dir.mkdir()
+
+    # Real Python next to the launcher, reachable through the shebang's
+    # "$(dirname "$REALPATH")<interp>" suffix — pass /python so it resolves
+    # to install_dir/python.
+    (install_dir / "python").symlink_to(sys.executable)
+
+    body = "import os\nprint(os.getcwd())\n"
+    launcher = install_dir / "cli"
+    launcher.write_text(format_shebang("/python", SHEBANG_TPL_MACOS) + body)
+    launcher.chmod(0o755)
+
+    # The Salt-on-macOS deployment shape: /usr/local/bin/<cmd> is a symlink
+    # to /opt/<app>/bin/<cmd>.  Reproduce that here.
+    shim = shim_dir / "cli"
+    shim.symlink_to(launcher)
+
+    result = subprocess.run(
+        [str(shim)],
+        cwd=str(caller_dir),
+        capture_output=True,
+        text=True,
+        check=False,
+    )
+    assert result.returncode == 0, f"stderr={result.stderr!r}"
+    reported_cwd = pathlib.Path(result.stdout.strip()).resolve()
+    assert reported_cwd == caller_dir.resolve(), (
+        f"launcher leaked cwd: got {reported_cwd}, expected {caller_dir.resolve()}"
+    )
+
+
 def test_relative_interpreter_default_location() -> None:
     assert relative_interpreter("/tmp/relenv", "/tmp/relenv/bin", "/tmp/relenv/bin/python3") == pathlib.Path(
         "..", "bin", "python3"

